pkg/executor/internal/mpp/executor_with_retry.go

```diff
@@ -0,0 +1,250 @@
+	// 2. If the held MPP results exceed the capacity, will starts returning results to caller.
+	//    Once the results start being returned, error recovery cannot be performed anymore.
+	mppErrRecovery *RecoveryHandler
+	planIDs        []int
+	KVRanges       []kv.KeyRange
+	queryID        kv.MPPQueryID
+	startTS        uint64
+	gatherID       uint64
+	nodeCnt        int
+}
+
+var _ kv.Response = &ExecutorWithRetry{}
+
+// NewExecutorWithRetry create ExecutorWithRetry.
+func NewExecutorWithRetry(ctx context.Context, sctx sessionctx.Context, parentTracker *memory.Tracker, planIDs []int,
+	plan plannercore.PhysicalPlan, startTS uint64, queryID kv.MPPQueryID,
+	dummy bool, is infoschema.InfoSchema) (*ExecutorWithRetry, error) {
+	// TODO: After add row info in tipb.DataPacket, we can use row count as capacity.
+	// For now, use the number of tipb.DataPacket as capacity.
+	const holdCap = 2
+
+	disaggTiFlashWithAutoScaler := config.GetGlobalConfig().DisaggregatedTiFlash && config.GetGlobalConfig().UseAutoScaler
+	_, allowTiFlashFallback := sctx.GetSessionVars().AllowFallbackToTiKV[kv.TiFlash]
+
+	// 1. For now, mpp err recovery only support MemLimit, which is only useful when AutoScaler is used.
+	// 2. When enable fallback to tikv, the returned mpp err will be ErrTiFlashServerTimeout,
+	//    which we cannot handle for now. Also there is no need to recovery because tikv will retry the query.
+	// 3. For cached table, will not dispatch tasks to TiFlash, so no need to recovery.
+	enableMPPRecovery := disaggTiFlashWithAutoScaler && !allowTiFlashFallback && !dummy
+
+	failpoint.Inject("mpp_recovery_test_mock_enable", func() {
+		if !dummy && !allowTiFlashFallback {
+			enableMPPRecovery = true
+		}
+	})
+
+	recoveryHandler := NewRecoveryHandler(disaggTiFlashWithAutoScaler,
+		uint64(holdCap), enableMPPRecovery, parentTracker)
+	retryer := &ExecutorWithRetry{
+		ctx:            ctx,
+		sctx:           sctx,
+		memTracker:     memory.NewTracker(parentTracker.Label(), 0),
+		planIDs:        planIDs,
+		is:             is,
+		plan:           plan,
+		startTS:        startTS,
+		queryID:        queryID,
+		mppErrRecovery: recoveryHandler,
+	}
+
+	var err error
+	retryer.KVRanges, err = retryer.setupMPPCoordinator(ctx, false)
+	return retryer, err
+}
+
+// Next implements kv.Response interface.
+func (r *ExecutorWithRetry) Next(ctx context.Context) (resp kv.ResultSubset, err error) {
+	if err = r.nextWithRecovery(ctx); err != nil {
+		return nil, err
+	}
+
+	if r.mppErrRecovery.NumHoldResp() != 0 {
+		if resp, err = r.mppErrRecovery.PopFrontResp(); err != nil {
+			return nil, err
+		}
+	} else if resp, err = r.coord.Next(ctx); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// Close implements kv.Response interface.
+func (r *ExecutorWithRetry) Close() error {
+	mppcoordmanager.InstanceMPPCoordinatorManager.Unregister(r.getCoordUniqueID())
+	// Handle reports only when it's not failed.
+	r.coord.(*localMppCoordinator).handleAllReports()
```

Comment:
I'm not sure if it is ok to call `handleAllReports()` before `coord.Close()`, maybe keep the `handleAllReports` inside `coord.Close()`, and add an extra argument to `coord.Close()` to indicate whether this `Close()` it should handleAllReports or not?

Reply:
Add a new method `closeWithoutReport` for `localMppCoordinator`.

Better not change `Close()`, because it implement `kv.Response` interface.